### DISCUSSION TITLE
Strange types returned when using Prisma rawQuery

### DISCUSSION THREAD

**Discussion by inhanbyeol** [prisma/prisma#24953] [Q&A] [answered]
### Question

If I use jsonb_build_object when writing a low query, will I lose the Date Type?
If so, why is the Date type guaranteed when I use jsonb_build_object, even if I include it in a Prisma findMany?

Furthermore, is there any way to get the date type guaranteed?

```ts
    async onModuleInit() {
        const test = await this.prisma
            .$queryRaw`SELECT c.id, c.value, c.created_at as "createdAt", jsonb_build_object('id', t.id, 'value', t.value, 'createdAt', t.created_at) as "postType" FROM "_Info:PostCategory" as c LEFT JOIN "_Info:PostType" as t ON c.post_type_id = t.id  WHERE c.post_type_id = 1`;
        console.log(test);
    }
```

```sql
SELECT c.id, c.value, c.created_at as "createdAt", jsonb_build_object('id', t.id, 'value', t.value, 'createdAt', t.created_at) as "postType" FROM "_Info:PostCategory" as c LEFT JOIN "_Info:PostType" as t ON c.post_type_id = t.id  WHERE c.post_type_id = 1
```

```ts
[
  {
    id: 1,
    value: '공지사항',
    createdAt: 2024-07-12T07:37:10.618Z,
    postType: { id: 1, value: '공지사항/이벤트', createdAt: '2024-07-12T07:37:10.555' }
  },
  {
    id: 2,
    value: '이벤트',
    createdAt: 2024-07-12T07:37:10.618Z,
    postType: { id: 1, value: '공지사항/이벤트', createdAt: '2024-07-12T07:37:10.555' }
  }
]

```

### Expected behavior (optional)

```ts
[
  {
    id: 1,
    value: '공지사항',
    createdAt: 2024-07-12T07:37:10.618Z,
    postType: { id: 1, value: '공지사항/이벤트', createdAt: 2024-07-12T07:37:10.555Z }
  },
  {
    id: 2,
    value: '이벤트',
    createdAt: 2024-07-12T07:37:10.618Z,
    postType: { id: 1, value: '공지사항/이벤트', createdAt: 2024-07-12T07:37:10.555Z }
  }
]

```

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
Hi @inhanbyeol 👋 

When using `jsonb_build_object` in a raw SQL query, you may encounter issues with the Date type being returned as a string. This is evident from your example where the createdAt field inside the postType object is returned as a string:
```
{
  id: 1,
  value: '공지사항',
  createdAt: 2024-07-12T07:37:10.618Z,
  postType: { id: 1, value: '공지사항/이벤트', createdAt: '2024-07-12T07:37:10.555' }
}
```
This behavior occurs because JSON does not have a native Date type, and when using `jsonb_build_object`, PostgreSQL converts the Date to a string.

> Why is the Date type guaranteed in Prisma findMany?

When using `findMany` method, Prisma handles the conversion of database types to JavaScript types, ensuring that DateTime fields are returned as JavaScript Date objects. This type safety is part of Prisma's ORM capabilities, which ensures that the types defined in your Prisma schema are respected and correctly mapped to JavaScript types.

To ensure that the Date type is preserved when using raw queries, you can manually parse the date strings into JavaScript Date objects after retrieving the data.
```js
async onModuleInit() {
    const result = await this.prisma.$queryRaw`
        SELECT 
            c.id, 
            c.value, 
            c.created_at as "createdAt", 
            jsonb_build_object('id', t.id, 'value', t.value, 'createdAt', t.created_at) as "postType" 
        FROM "_Info:PostCategory" as c 
        LEFT JOIN "_Info:PostType" as t 
        ON c.post_type_id = t.id  
        WHERE c.post_type_id = 1
    `;

    const parsedResult = result.map(item => ({
        ...item,
        createdAt: new Date(item.createdAt),
        postType: {
            ...item.postType,
            createdAt: new Date(item.postType.createdAt)
        }
    }));

    console.log(parsedResult);
}
```

**Reply by inhanbyeol** [reply]
Thank you for your response 
So if there is an issue with Date type being returned as a string in jsonb, is it only Date type? Int type seems to be fine, but boolean and Enum?

**Reply by ludralph** [reply]
For other types like Int, Boolean, and Enum, the behavior is generally more straightforward:

- Int: This type is typically serialized and deserialized as a number in JSON.
- Boolean: This type should be serialized and deserialized correctly as a boolean value (true or false) in JSON.
- Enum: Enum values are usually serialized as strings representing the enum values.
